[Shipping labels] Add green print label card to shipments

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Part of WOOMOB-284

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a green “Print Label” card to each shipment tab. After purchasing a new label, the “Print Label” card will be displayed on the current screen instead of opening a new one.

The PR may appear large, but it primarily copies existing code (from WooShippingLabelPurchasedViewModel to WooShippingLabelCreationViewModel). In a follow-up PR, I’ll remove redundant classes related to WooShippingLabelPurchasedFragment and consider adding test cases for the new logic.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
1. Go to the Orders tab and select an order.
2. Tap an order with shipping labels.
3. Tap on Create shipping label.
4. Tap the Split shipment button.
5. Create new shipments.
6. Tap the DONE button.
7. Purchase one of the shipments.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
[Screen_recording_20250603_032139.webm](https://github.com/user-attachments/assets/04d58ad8-33c9-4b8b-b6b8-7f3161b69b6d)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add purchased shipment events to WooShippingLabelCreationFragment

irfano committed:
Remove redundant 'suspend' modifier from ObserveShippingLabelStatus

irfano committed:
Use static selectedShipmentIndex for purchasing shipments functions

irfano committed:
Fix incorrect format for the response of config endpoint

irfano committed:
Use equality check instead of elvis for nullable boolean check

irfano committed:
Migrate `purchaseState` to `ShipmentUIModel`

This commit moves the `purchaseState` property from `WooShippingLabelCreationViewModel` to `ShipmentUIModel`. This change allows each shipment to have its own independent purchase state, providing more granular control and better reflecting the reality of multi-shipment scenarios.

irfano committed:
Remove unused combine function with 10 flows

irfano committed:
Populate label data in shipments

This commit modifies the `GetShipments` use case to populate the `purchased`, `labelId`, `carrierId`, and `trackingNumber` fields in the `ShipmentUIModel` if corresponding label data exists.

irfano committed:
Use a different response model for the purchase label endpoint

Previously, the response included an id parameter in number format, but it actually returned a string, causing a failure.

irfano committed:
Remove WooShippingLabelPurchasedFragment references

This is preparation to remove WooShippingLabelPurchasedFragment completely.

irfano committed:
Handle purchase in creation screen instead of opening a new fragment

irfano committed:
Fix detekt error

irfano committed:
Sort Woo Shipping shipments by ID

irfano committed:
Move part of WooShippingLabelPurchasedScreen to a separate composable

irfano committed:
Use `status` instead of `isPurchaseFinished` in WooShippingLabelCreationViewModel

irfano committed:
Rename ShippingLabelStatus enum values to uppercase

This commit updates the `ShippingLabelStatus` enum values to follow the standard Java/Kotlin enum naming convention (uppercase with underscores).

irfano committed:
Remove manual mapping for `ShippingLabelStatus`

The `ShippingLabelStatus` field in `PurchasedLabelDTO` and `ShippingLabelDTO` will now be directly deserialized from the JSON response using a custom `JsonDeserializer`. This removes the need for the `mapShippingLabelStatus` function in `WooShippingNetworkingMapper` and associated string constants.

irfano committed:
Remove unnecessary unit tests

irfano committed:
Add PrintShippingLabelSection composable to the shipment creation screen

irfano committed:
Add PrintShippingLabelSection events into the view model

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/GetShipments.kt:
```diff
@@ -54,9 +55,31 @@ class GetShipments @Inject constructor(
-                ShipmentUIModel(id = shipmentId, remoteId = shipmentId, items = items, purchased = purchased)
+                ShipmentUIModel(id = shipmentId, remoteId = shipmentId, items = items)
             }
+        }.sortedBy { it.id.toLong() }
```
I added sorting to align with iOS and web behavior.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationFragment.kt:
```diff
@@ -153,4 +154,28 @@ class WooShippingLabelCreationFragment : BaseFragment(), BackPressListener {
 
     override fun onRequestAllowBackPress(): Boolean = viewModel.allowBackNavigation()
+
+    private fun openShippingLabelPreview(file: File) {
+        ActivityUtils.previewPDFFile(requireActivity(), file)
+    }
+
+    private fun openLearnMoreView() {
+        findNavController().navigate(
+            WooShippingLabelCreationFragmentDirections
+                .actionWooShippingLabelCreationFragmentToPrintShippingLabelInfoFragment()
+        )
+    }
+
+    private fun openUrl(url: String) {
+        ChromeCustomTabUtils.launchUrl(requireContext(), url)
+    }
+
+    private fun showErrorDialog(messageResId: Int) {
+        WooDialog.showDialog(
+            activity = requireActivity(),
+            titleId = R.string.error_generic,
+            messageId = messageResId,
+            positiveButtonId = R.string.dialog_ok
+        )
+    }
```
These were copied from `WooShippingLabelPurchasedFragment`

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -643,37 +673,22 @@ class WooShippingLabelCreationViewModel @Inject constructor(
-                        shippingLines = currentViewState.shippingLines,
-                        hazmatSelection = hazmatSelection
-                    ).let {
-                        triggerEvent(
-                            LabelPurchased(
-                                purchaseData = it,
-                                totalItems = currentViewState.totalItems,
-                                totalItemsCost = currentViewState.totalItemsCost
-                            )
-                        )
```
We won't be opening `WooShippingLabelPurchasedFragment` anymore. Instead, we will show the purchased state on the current screen.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -969,6 +1018,29 @@ class WooShippingLabelCreationViewModel @Inject constructor(
             get() = (this as? Declared)?.hazmatCategory
     }
 
+    data class OpenShippingLabelFile(val file: File) : Event()
+    data class OpenUrl(val url: String) : Event()
+    data class ShowError(val errorResId: Int) : Event()
+    object StartRefundRequest : Event()
+    object OpenLearnMoreScreen : Event()
+
+    enum class Carrier(val pickupUrl: String) {
+        USPS("https://tools.usps.com/schedule-pickup-steps.htm"),
+        UPS("https://wwwapps.ups.com/pickup/request"),
+        DHL("https://mydhl.express.dhl/us/en/schedule-pickup.html#/schedule-pickup#label-reference");
+
+        companion object {
+            fun fromCarrierId(carrierId: String): Carrier? {
+                return when (carrierId) {
+                    "usps" -> USPS
+                    "ups" -> UPS
+                    "dhlexpress" -> DHL
+                    else -> null
+                }
+            }
+        }
+    }
+
```
These were copied from `WooShippingLabelPurchasedViewModel`

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/components/PrintShippingLabelSection.kt:
This file is mostly the same as `WooShippingLabelPurchasedScreen`.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/DTOs.kt:
```diff
@@ -74,20 +98,18 @@ data class ShippingLabelDTO(
     @SerializedName("is_letter") val isLetter: Boolean? = null,
     @SerializedName("product_names") val productNames: List<String>? = null,
     @SerializedName("product_ids") val productIds: List<Long>? = null,
+    @SerializedName("id") val shipmentId: Long? = null,
```
The new id parameter is sometimes a number and sometimes represents a package name (e.g., “small_box”) on the backend. That’s why I created a separate `PurchasedLabelDTO`.

hichamboushaba replied:
I'm not following here, can you explain more please? 🙏

irfano replied:
Previously, `ShippingLabelDTO` was used for both the config and purchase endpoints. We were using the same class to handle responses from both, but there’s a key difference: the config endpoint returns the id as an integer (e.g., id=10), whereas the purchase endpoint returns it as a string (e.g., id="small_box"). This caused a crash because we were expecting an int, but received a String.

I’m aware there are alternative solutions, such as using a custom deserializer, but I chose to separate the response classes in this case. Let me know if you’re aware of a simpler solution.

Feel free to ask for more details, I can share example responses from the endpoints if needed.

hichamboushaba replied:
I see, this confused me because I don't see an `id` field in the `PurchasedLabelDTO`, but that's fine if we don't need it.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/DTOs.kt:
```diff
@@ -184,3 +206,17 @@ private class ShipmentMapDeserializer : JsonDeserializer<ShipmentMap> {
     }
 }
+
+private class ShippingLabelStatusDeserializer : JsonDeserializer<ShippingLabelStatus> {
+    override fun deserialize(
+        json: JsonElement,
+        typeOfT: Type,
+        context: JsonDeserializationContext
+    ) = when (json.asString.uppercase()) {
+        "PURCHASE_IN_PROGRESS" -> ShippingLabelStatus.PURCHASE_IN_PROGRESS
+        "PURCHASED" -> ShippingLabelStatus.PURCHASED
+        "PURCHASE_ERROR" -> ShippingLabelStatus.PURCHASE_ERROR
+        "ANONYMIZED" -> ShippingLabelStatus.ANONYMIZED
+        else -> ShippingLabelStatus.UNKNOWN
+    }
+}
```
I added a JSON deserializer for `status`. Previously, it used a manual mapper, but handling this at the JSON parsing layer is easier.

hichamboushaba replied:
That's definitely better than the manual approach used before, but I think Gson supports by default the same logic, it can deserialize enums without any added logic when the names match the enum values (as shared [here](https://stackoverflow.com/a/16749837) too), WDYT?

irfano replied:
I believe using hardcoded values from the endpoint response is safer here. Otherwise, these enums might be renamed in the future by someone (possibly me again), who may overlook the fact that they directly represent the backend response. WDYT?

Alternatively, we could create a separate enum specifically for the networking layer and use a different one for the UI layer, but I think sticking with hardcoded values is simpler.

hichamboushaba replied:
If we are worried about renaming the enum values, we can use `@SerializedName` to force the names.

irfano replied:
Thanks for the suggestion. I made the related change and tested it, it works as expected. Using enums makes it cleaner: 9fd0126513bdd7818ce00be2b524d8ddf3e1102d

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapper.kt:
```diff
@@ -39,17 +39,18 @@ class WooShippingNetworkingMapper @Inject constructor(
             commercialInvoiceUrl = shippingLabelDTO.commercialInvoiceUrl.orEmpty(),
-            isCommercialInvoiceSubmittedElectronically =
-            shippingLabelDTO.isCommercialInvoiceSubmittedElectronically ?: false,
+            isCommercialInvoiceSubmittedElectronically = shippingLabelDTO
+                .isCommercialInvoiceSubmittedElectronically == true,
```
These `?: false` → `== true` change was suggested by IDE.

irfano commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModelTest.kt:
```diff
@@ -764,60 +728,6 @@ class WooShippingLabelCreationViewModelTest : BaseUnitTest() {
         }
 
-    @Test
-    fun `when onPurchaseShippingLabel succeed then return the label data`() = testBlocking {
```
This test case is not valid anymore.

irfano commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapperTest.kt:
```diff
@@ -44,38 +42,4 @@ class WooShippingNetworkingMapperTest : BaseUnitTest() {
         assertEquals(result.originCountry, dto.originCountry)
     }
-
-    @Test
-    fun `when a shipping label dto is received then return the expected model`() {
-        val dto = ShippingLabelDTO(
-            labelId = 1234,
-            tracking = "123456789",
-            refundableAmount = BigDecimal.ZERO,
-            status = "PURCHASED"
-        )
-
-        val result = sut.invoke(dto)
-
-        assertEquals(result.labelId, dto.labelId)
-        assertEquals(result.tracking, dto.tracking)
-        assertEquals(result.refundableAmount, dto.refundableAmount)
-        assertEquals(result.status, ShippingLabelStatus.Purchased)
-    }
-
-    @Test
-    fun `when a shipping label dto is received with an unknown status then return the expected model`() {
-        val dto = ShippingLabelDTO(
-            labelId = 1234,
-            tracking = "123456789",
-            refundableAmount = BigDecimal.ZERO,
-            status = "NOT_VALID"
-        )
-
-        val result = sut.invoke(dto)
-
-        assertEquals(result.labelId, dto.labelId)
-        assertEquals(result.tracking, dto.tracking)
-        assertEquals(result.refundableAmount, dto.refundableAmount)
-        assertEquals(result.status, ShippingLabelStatus.Unknown)
-    }
```
I removed these test cases since they no longer test any meaningful logic. They might have been relevant when manual status mapping existed, but that’s no longer the case.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/DTOs.kt:
```diff
@@ -44,7 +45,11 @@ data class ConfigDTO(
     @SerializedName("shipments")
     @JsonAdapter(ShipmentMapDeserializer::class)
     val shipments: ShipmentMap? = null,
-    @SerializedName("shippingLabelData") val shippingLabelData: ShippingLabelDTO,
+    @SerializedName("shippingLabelData") val shippingLabelData: ShippingLabelDataDTO,
```
The data format for `ConfigDTO` was incorrect before. I’ve fixed it here.

hichamboushaba replied:
Good catch

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/GetShipments.kt:
`GetShipments` retrieves existing shipments to display shipment tabs. Updating this use case was necessary because previously we only showed the purchased state after buying a label. Now, we need to display it for each shipment when the shipping labels screen is opened.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/purchased/WooShippingLabelPurchasedFragment.kt:
This class will be removed later.

irfano committed:
Adjust paddings on the Create shipping label screen

irfano committed:
Update default paper size option to Label

irfano committed:
Remove unused classes related to shipping label purchased screen

irfano committed:
Add tests for cases after moving print card into label creation screen

irfano committed:
Rename `id` to `localId` in `ShipmentUIModel`

hichamboushaba reviewed (approved):
Great work @irfano, this works well, and the code looks good.

I added some comments, but none is blocking, and I noticed a minor issue: the screen title changes from "Print shipping label" to "Create shipping label" after tapping on "print shipping label" (during the loading step).
I'm pre-approving.

irfano committed:
[Shipping Labels] Display correct title on loading screen

irfano commented:
> the screen title changes from "Print shipping label" to "Create shipping label" after tapping on "print shipping label" (during the loading step).

Great catch! I fixed it, but in a wrong PR. 🤦🏻‍♂️ You can test it in #14142

irfano committed:
Use SerializedName for ShippingLabelStatus enum

This commit updates the `ShippingLabelStatus` enum to use `@SerializedName` annotations for its values, ensuring correct serialization and deserialization with Gson.

The custom `ShippingLabelStatusDeserializer` has been removed as it is no longer necessary with the use of `@SerializedName`.

irfano committed:
Merge pull request #14142 from woocommerce/issue/WOOMOB-284-clear-purchased-label-screen-add-unit-tests

[Shipping labels] Clear purchased label screen and add unit tests

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-284-add-green-print-label-card-to-shipments

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-284-add-green-print-label-card-to-shipments
